### PR TITLE
DE-35424 Add ExecutionPlannedAt to the JobInterface

### PR DESCRIPTION
**Pull request by jakubkulhanek**
In this PR we are solving pushing delay jobs to the SQS with delay over SQS delay limit.
How it is work:
When a job has delay bigger then 15 minutes (for example 30 minutes) we set executionPlannedAt to current datetime plus 30 minutes. Before this job is executed again then check the remains time from  executionPlannedAt and current datetime and when the diff is bigger then 0  we delay this job again until the diff between executionPlannedAt and current datetime is 0 then the job is processed

### CODE REVIEW

File: src/Jobs/JobInterface.php [hunk 1]
```diff
@@ -47,4 +47,10 @@ public function incrementAttempts(): void;
 
 
     public function getJobDefinition(): JobDefinitionInterface;
+
+
+    public function getTimeOfExecution(): ?int;
+
+
+    public function setTimeOfExecution(?int $timeOfExecution): void;  // $timeOfExecution = Timestamp of execution time
```
Comment (dominikvoda) [inline src/Jobs/JobInterface.php:55]: name the method in a way we don't need extra comment

**Review by dominikvoda — CHANGES_REQUESTED** [review]
missing tests

File: src/Jobs/JobInterface.php [hunk 2]
```diff
@@ -47,4 +47,10 @@ public function incrementAttempts(): void;
 
 
     public function getJobDefinition(): JobDefinitionInterface;
+
+
+    public function getTimeOfExecution(): ?int;
+
+
+    public function setTimeOfExecution(?int $timeOfExecution): void;
```
Comment (PetrHeinz) [inline src/Jobs/JobInterface.php:55]: Why `int` and not `DateTimeImmutable`?

Maybe it would deserve a better naming. How about `getExecutionPlannedAt(): ?DateTimeImmutable` and `executionPlanned(?DateTimeImmutable $plannedAt)`, to be a bit more concise with `getExecutionStartedAt` / `executionStarted`.

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 3]
```diff
@@ -213,6 +213,11 @@ public function pushDelayed(JobInterface $job, int $delayInSeconds): void
             self::DELAY_SECONDS => $delayInSeconds,
```
Comment (PetrHeinz) [inline src/Queue/AWSSQS/SqsQueueManager.php:213]: Here you still delay it by `$delayInSeconds` and not use the new attribute. Maybe you should add a `TODO` if you're planning to add it in a new PR...

File: src/Jobs/FailResolving/PushDelayedResolver.php [hunk 4]
```diff
@@ -27,6 +27,7 @@ public function __construct(QueueManagerInterface $queueManager, LoggerInterface
     public function resolve(JobInterface $job, Throwable $exception): void
     {
         $job->incrementAttempts();
+        $job->setTimeOfExecution(null);
```
Comment (PetrHeinz) [inline src/Jobs/FailResolving/PushDelayedResolver.php:30]: This seems a bit unintuitive. Why do we set that to NULL during resolving? Maybe we should name it better or unify it with `incrementAttempts`...

File: src/Jobs/JobInterface.php [hunk 5]
```diff
@@ -47,4 +47,10 @@ public function incrementAttempts(): void;
 
 
     public function getJobDefinition(): JobDefinitionInterface;
+
+
+    public function getTimeOfExecution(): ?int;
```
Comment (PetrHeinz) [inline src/Jobs/JobInterface.php:52]: This value is not used anywhere (except as a hasser). Maybe a TODO is missing?

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 6]
```diff
@@ -213,6 +213,11 @@ public function pushDelayed(JobInterface $job, int $delayInSeconds): void
...
         ];
 
+        $timeOfExecution = $job->getTimeOfExecution();
+        if ($timeOfExecution === null) {
+            $job->setTimeOfExecution($job->getCreatedAt()->getTimestamp() + $delayInSeconds);
```
Comment (PetrHeinz) [inline src/Queue/AWSSQS/SqsQueueManager.php:218]: Why is this set only in the SQS queue manager? In RabbitMQ, all job will have execution time of NULL?

**Review by PetrHeinz — CHANGES_REQUESTED** [review]
PR description is missing. This PR just adds a new attribute and it's not really clear how it's planned to be used...

If this work is done here in the library, it should be really intuitive to use. Maybe we can implement this feature in our code instead, if that's too hard to achieve?

Also, this will create a BC break (as you're adding new methods to an existing interface), it should be clearly marked as such.

File: src/Jobs/JobInterface.php (hunk 2 above)
Comment (jakubkulhanek) [reply src/Jobs/JobInterface.php:55]: I took this name from platform. For me it is better to use simillary naming

File: src/Queue/AWSSQS/SqsQueueManager.php (hunk 3 above)
Comment (jakubkulhanek) [reply src/Queue/AWSSQS/SqsQueueManager.php:213]: I do not understand what you mean

File: src/Queue/AWSSQS/SqsQueueManager.php (hunk 6 above)
Comment (jakubkulhanek) [reply src/Queue/AWSSQS/SqsQueueManager.php:218]: Because this delaying must work only for SQS

File: src/Jobs/JobInterface.php (hunk 5 above)
Comment (jakubkulhanek) [reply src/Jobs/JobInterface.php:52]: ?

File: src/Jobs/FailResolving/PushDelayedResolver.php (hunk 4 above)
Comment (jakubkulhanek) [reply src/Jobs/FailResolving/PushDelayedResolver.php:30]: ?

File: tests/Queue/AWSSQS/SqsConsumerTest.php [hunk 7]
```diff
@@ -207,6 +220,58 @@ public function testRequeueDelayableProcessFailWarningOnly(): void
     }
 
 
+    public function testDelayJobWithExecutionPlannedAt(): void
+    {
+        $exampleJob = new ExampleJob();
+        $exampleJob->setExecutionPlannedAt(new DateTimeImmutable('2016-08-15T17:00:00+00:00'));
```
Comment (PetrHeinz) [inline tests/Queue/AWSSQS/SqsConsumerTest.php:226]: [o] It would be nice to have at least two different data sets, one in the future (just like you have) and one in immediate future (`2016-08-15T15:00:01+00:00`) to test the edge case. Ideally, you should also test time zones.

File: tests/Queue/AWSSQS/SqsConsumerTest.php [hunk 8]
```diff
@@ -207,6 +220,58 @@ public function testRequeueDelayableProcessFailWarningOnly(): void
...
+        $sqsMessage = $this->createSqsMessage($this->getSqsMessageData());
+        $sqsConsumer = $this->createSqsConsumer($this->sqsClientMock);
+        $sqsConsumer($sqsMessage);
+    }
+
+
+    public function testExecuteJobWithExecutionPlannedAt(): void
+    {
+        $exampleJob = new ExampleJob();
+        $exampleJob->setExecutionPlannedAt(new DateTimeImmutable('2016-08-15T14:00:00+00:00'));
```
Comment (PetrHeinz) [inline tests/Queue/AWSSQS/SqsConsumerTest.php:254]: [o] It would be nice to have at least two different data sets, one in the past (just like you have) and one in the present (`2016-08-15T15:00:00+00:00`) to test the edge case. Ideally, you should also test time zones.

**Review by PetrHeinz — APPROVED** [review]
PR description is still missing.

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 9]
```diff
@@ -209,11 +216,21 @@ public function pushDelayed(JobInterface $job, int $delayInSeconds): void
...
 
-        $parameters = [
-            self::DELAY_SECONDS => $delayInSeconds,
-        ];
+        if ($delayInSeconds > self::MAX_DELAY_SECONDS) {
+            $executionPlannedAt = $this->dateTimeImmutableFactory->getNow()->modify(
+                sprintf('+ %d seconds', $delayInSeconds),
+            );
+            $this->logger->info(
+                'Requested delays is greater than SQS limit. Job execution has been planned and will be requeued until then.',
```
Comment (PetrHeinz) [inline src/Queue/AWSSQS/SqsQueueManager.php:224]: type (probably mine :grin:)
```suggestion
                'Requested delay is greater than SQS limit. Job execution has been planned and will be requeued until then.',
```

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 10]
```diff
@@ -209,11 +216,21 @@ public function pushDelayed(JobInterface $job, int $delayInSeconds): void
...
 
-        $parameters = [
-            self::DELAY_SECONDS => $delayInSeconds,
-        ];
+        if ($delayInSeconds > self::MAX_DELAY_SECONDS) {
+            $executionPlannedAt = $this->dateTimeImmutableFactory->getNow()->modify(
+                sprintf('+ %d seconds', $delayInSeconds),
+            );
+            $this->logger->info(
+                'Requested delas is greater than SQS limit. Job execution has been planned and will be requeued until then.',
```
Comment (PetrHeinz) [inline src/Queue/AWSSQS/SqsQueueManager.php:224]: You deleted the wrong letter :rofl: 
```suggestion
                'Requested delay is greater than SQS limit. Job execution has been planned and will be requeued until then.',
```

File: tests/Jobs/Execution/JobLoaderTest.php [hunk 11]
```diff
@@ -81,8 +92,14 @@ public function testLoadSimpleJob(): void
...
-        Assert::assertNotNull($simpleJob->getExecutionPlannedAt());
-        DateTimeAssertions::assertDateTimeAtomEqualsDateTime(ExampleJob::EXECUTION_PLANNED_AT, $simpleJob->getExecutionPlannedAt());
+
+        if ($executionPlannedAt !== null) {
+            Assert::assertNotNull($simpleJob->getExecutionPlannedAt());
+        }
+
+        if ($executionPlannedAt === null) {
+            Assert::assertNull($simpleJob->getExecutionPlannedAt());
+        }
```
Comment (PetrHeinz) [inline tests/Jobs/Execution/JobLoaderTest.php:102]: `Assert::assertEquals()` tady nefunguje?

File: tests/Queue/AWSSQS/SqsQueueManagerTest.php [hunk 12]
```diff
@@ -214,7 +214,7 @@ public function testPushDelayedWithJobDelayOverSqsMaxDelayLimit(string $queueNam
 
         $queueManager->pushDelayed($exampleJob, 1800);
 
-        $this->loggerMock->hasInfo('Requested delays is greater than SQS limit. Job execution has been planned and will be requeued until then.');
+        $this->loggerMock->hasInfo('Requested delay is greater than SQS limit. Job execution has been planned and will be requeued until then.');
```
Comment (PetrHeinz) [inline tests/Queue/AWSSQS/SqsQueueManagerTest.php:217]: Tohle není assert, jen zavoláš hasser, který ti vrátí FALSE (zrovna teď, protože je v implementaci typo s `delas`).

File: src/Jobs/HasExecutionPlannedAt.php [hunk 13]
```diff
@@ -0,0 +1,22 @@
+<?php declare(strict_types = 1);
+
+namespace BE\QueueManagement\Jobs;
+
+use DateTimeImmutable;
+
+trait HasExecutionPlannedAt
```
Comment (lukasHod) [inline src/Jobs/HasExecutionPlannedAt.php:7]: what is the reason having this trait? We can just extend SimpleJob with those methods. Is there any plan to use it somewhere else?

Comment (jakubkulhanek) [reply src/Jobs/HasExecutionPlannedAt.php:7]: It is for easier work in jobs which are using the JobInterface.
It is good to have one implementation in trait instead of write same code in different jobs

**Review by lukasHod — APPROVED** [review]
